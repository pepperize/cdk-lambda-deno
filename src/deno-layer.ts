import * as path from "path";
import { Stack } from "aws-cdk-lib";
import * as lambda from "aws-cdk-lib/aws-lambda";
import { Construct } from "constructs";

export class DenoLayer extends lambda.LayerVersion {
  public static getOrCreate(scope: Construct): DenoLayer {
    const stack = Stack.of(scope);
    const id = "DenoLayer";
    const existing = stack.node.tryFindChild(id);
    return (existing as DenoLayer) || new DenoLayer(stack, id);
  }

  public constructor(scope: Construct, id: string) {
    const props: lambda.LayerVersionProps = {
      code: lambda.Code.fromAsset(path.join(__dirname, "layer.zip")),
      description: "/bin/deno",
    };

    super(scope, id, props);
  }
}
