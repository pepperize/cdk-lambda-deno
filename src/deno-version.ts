// https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
const regex =
  /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/g;

export class DenoVersion {
  /**
   * 1.23.3
   */
  public static readonly V1_23_3 = new DenoVersion("1.23.3");

  /**
   * Specify a custom version.
   * Use this if the version you need is not available in one of the predefined versions.
   *
   * @param version The semantic version number without prefix 'v'.
   */
  public static of(version: string): DenoVersion {
    if (!regex.test(version)) {
      throw new Error(`'${version}' is not a valid semantic version`);
    }

    return new DenoVersion(version);
  }

  /**
   * The version string.
   */
  public readonly version: string;

  private constructor(version: string) {
    this.version = version;
  }
}
