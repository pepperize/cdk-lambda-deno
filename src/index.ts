export * from "./deno-function";
export * from "./deno-layer";
