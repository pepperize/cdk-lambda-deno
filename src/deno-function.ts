import * as lambda from "aws-cdk-lib/aws-lambda";
import { Construct } from "constructs";
import { DenoLayer } from "./deno-layer";

export interface DenoFunctionProps extends lambda.FunctionOptions {
  readonly code: lambda.Code;
  readonly handler?: string;
}

export class DenoFunction extends lambda.Function {
  public constructor(scope: Construct, id: string, props: DenoFunctionProps) {
    const layer: lambda.ILayerVersion = DenoLayer.getOrCreate(scope);
    const layers = props.layers || [];
    layers.unshift(layer);
    const handler = props.handler ?? "index.handler";

    super(scope, id, {
      runtime: lambda.Runtime.PROVIDED_AL2,
      handler: handler,
      ...props,
      layers: layers,
    });
  }
}
