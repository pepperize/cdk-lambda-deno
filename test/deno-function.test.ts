import * as path from "path";
import { Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import * as lambda from "aws-cdk-lib/aws-lambda";
import { DenoFunction } from "../src";

describe("DenoFunction", () => {
  it("should match snapshot", () => {
    // Given
    const stack = new Stack();

    // When
    new DenoFunction(stack, "Function", {
      code: lambda.Code.fromAsset(path.join(__dirname, "../example/function")),
    });

    // Then
    const template = Template.fromStack(stack);

    expect(template.toJSON()).toMatchSnapshot();
  });
  it("should synthesize to a function", () => {
    // Given
    const stack = new Stack();

    // When
    new DenoFunction(stack, "Function", {
      code: lambda.Code.fromAsset(path.join(__dirname, "../example/function")),
    });

    // Then
    Template.fromStack(stack).hasResourceProperties("AWS::Lambda::Function", {
      Handler: "index.handler",
    });
  });
});
