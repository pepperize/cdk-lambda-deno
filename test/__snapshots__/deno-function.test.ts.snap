// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DenoFunction should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "DenoLayerD97F84F7": Object {
      "Properties": Object {
        "Content": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "bf204e6da0c6c1f29bcde3f7cddbc0664e5d97527a54c24989535bf0ab83f541.zip",
        },
        "Description": "/bin/deno",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "Function76856677": Object {
      "DependsOn": Array [
        "FunctionServiceRole675BB04A",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "a24e3ac5f2ab214690afdbea5578e5089f7d2ede5e57dc64b925eeb69516a3a6.zip",
        },
        "Handler": "index.handler",
        "Layers": Array [
          Object {
            "Ref": "DenoLayerD97F84F7",
          },
        ],
        "Role": Object {
          "Fn::GetAtt": Array [
            "FunctionServiceRole675BB04A",
            "Arn",
          ],
        },
        "Runtime": "provided.al2",
      },
      "Type": "AWS::Lambda::Function",
    },
    "FunctionServiceRole675BB04A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
